File: skills/inspect-and-debug/harness-doctor/scripts/lib/aggregate-commands.ts
/**
 * The `commands` and `errors` sections of the audit.
 *
 * These two carry most of the diagnostic value: `flagFlailing` and `retryLoops`
 * are what catch an agent groping for the right invocation, so the windowing is
 * spelled out rather than approximated.
 */

import {
  bump,
  byTime,
  clip,
  cmpStr,
  collapseWs,
  groupBy,
  median,
  percentile,
  ranking,
  safeDiv,
  round,
  sum,
  topKey,
  uniqSorted,
} from "./stats.ts";
import type {
  BashRow,
  CommandFamilyRow,
  CommandsSection,
  ErrorClusterRow,
  ErrorRow,
  ErrorsSection,
  FlagFlailingRow,
  MissingBinaryRow,
  PermissionDeniedRow,
  RepeatCommandRow,
  RetryLoopRow,
} from "./types.ts";

/** Seconds a burst of same-bin commands may span and still count as one flail. */
const FLAIL_WINDOW_SEC = 600;

/** Distinct normalised commands inside one window that make it a flail. */
const FLAIL_MIN_DISTINCT = 4;

/** Seconds after a failure inside which a re-run counts as a retry. */
const RETRY_WINDOW_SEC = 300;

/** Duration of a bash call in seconds, 0 when the transcript had no result time. */
const secOf = (row: { durSec?: number }) => row.durSec ?? 0;

/** Comparable form of a command: whitespace collapsed, nothing else changed. */
const cmdKey = (cmd: string) => collapseWs(cmd);

/** Roll a set of bash rows sharing a family into one ranked row. */
const familyRow = (family: string, rows: readonly BashRow[]): CommandFamilyRow => {
  const secs = rows.map(secOf);
  const bins = new Map<string, number>();
  for (const row of rows) bump(bins, row.bin);
  return {
    family,
    bin: topKey(bins),
    count: rows.length,
    totalSec: round(sum(secs, (s) => s), 1),
    medianSec: round(median(secs), 2),
    p95Sec: round(percentile(secs, 0.95), 2),
    errRate: round(safeDiv(rows.filter((r) => r.err).length, rows.length), 4),
  };
};

/**
 * Command families ranked by seconds and by count, for a subset of bash rows.
 * Exported so the projects section can reuse the same maths per project.
 */
export const commandFamilies = (bash: readonly BashRow[]) =>
  [...groupBy(bash, (r) => r.family)]
    .map(([family, rows]) => familyRow(family, rows))
    .sort((a, b) => cmpStr(a.family, b.family));

/** Identical command re-run inside one session: everything after the first run. */
const repeatRows = (bash: readonly BashRow[], top: number) => {
  const perCmd = new Map<
    string,
    { family: string; cmd: string; occurrences: number; sessions: Set<string>; wastedSec: number }
  >();
  for (const [, rows] of groupBy(bash, (r) => `${r.sid} ${cmdKey(r.cmd)}`)) {
    if (rows.length < 2) continue;
    const ordered = [...rows].sort((a, b) => (a.t ?? 0) - (b.t ?? 0));
    const first = ordered[0];
    if (!first) continue;
    const key = cmdKey(first.cmd);
    const entry = perCmd.get(key) ?? {
      family: first.family,
      cmd: clip(key, 200),
      occurrences: 0,
      sessions: new Set<string>(),
      wastedSec: 0,
    };
    entry.occurrences += ordered.length - 1;
    entry.sessions.add(first.sid);
    entry.wastedSec += sum(ordered.slice(1), secOf);
    perCmd.set(key, entry);
  }
  const rows: RepeatCommandRow[] = [...perCmd.values()].map((e) => ({
    family: e.family,
    cmd: e.cmd,
    occurrences: e.occurrences,
    sessions: e.sessions.size,
    wastedSec: round(e.wastedSec, 1),
  }));
  return ranking(rows, (r) => r.occurrences, (r) => r.cmd, top);
};

/** One burst of same-bin commands that all started inside the flail window. */
interface Burst {
  rows: BashRow[];
  families: Set<string>;
}

/**
 * Split time-ordered rows into bursts: a burst holds every row that starts within
 * FLAIL_WINDOW_SEC of the row that opened it, and the first row outside opens the
 * next burst. Greedy and left-to-right, so the split never depends on row order
 * beyond time.
 */
const burstsOf = (rows: readonly BashRow[]) => {
  const out: Burst[] = [];
  let current: Burst | undefined;
  let openedAt = 0;
  for (const row of rows) {
    const t = row.t ?? 0;
    if (!current || t - openedAt > FLAIL_WINDOW_SEC) {
      current = { rows: [], families: new Set<string>() };
      openedAt = t;
      out.push(current);
    }
    current.rows.push(row);
    current.families.add(row.family);
  }
  return out;
};

/**
 * Same `bin`, four or more distinct normalised commands inside ten minutes of one
 * session: the shape of an agent guessing at flags.
 */
const flailingRows = (bash: readonly BashRow[], top: number) => {
  const perBin = new Map<
    string,
    { sessions: Set<string>; occurrences: number; totalSec: number; cmds: Set<string> }
  >();
  for (const [key, rows] of groupBy(bash, (r) => `${r.sid} ${r.bin}`)) {
    const bin = key.slice(key.indexOf(" ") + 1);
    const sid = key.slice(0, key.indexOf(" "));
    for (const burst of burstsOf(byTime(rows))) {
      if (burst.families.size < FLAIL_MIN_DISTINCT) continue;
      const entry = perBin.get(bin) ?? {
        sessions: new Set<string>(),
        occurrences: 0,
        totalSec: 0,
        cmds: new Set<string>(),
      };
      entry.sessions.add(sid);
      entry.occurrences += burst.families.size;
      entry.totalSec += sum(burst.rows, secOf);
      for (const row of burst.rows) entry.cmds.add(clip(cmdKey(row.cmd), 160));
      perBin.set(bin, entry);
    }
  }
  const rows: FlagFlailingRow[] = [...perBin].map(([bin, e]) => ({
    bin,
    sessions: e.sessions.size,
    occurrences: e.occurrences,
    totalSec: round(e.totalSec, 1),
    sampleCmds: uniqSorted(e.cmds).slice(0, 4),
  }));
  return ranking(rows, (r) => r.occurrences, (r) => r.bin, top);
};

/**
 * Bash cost, ranked and diagnosed: families by seconds and by count, commands
 * re-run inside one session, and bins the agent flailed on.
 */
export const commandsSection = (
  bash: readonly BashRow[],
  top: number,
): CommandsSection => {
  const families = commandFamilies(bash);
  return {
    byTotalSec: ranking(families, (r) => r.totalSec, (r) => r.family, top),
    byCount: ranking(families, (r) => r.count, (r) => r.family, top),
    repeatsInSession: repeatRows(bash, top),
    flagFlailing: flailingRows(bash, top),
  };
};

/** Patterns that name a binary the shell could not find, most specific first. */
const MISSING_BIN_PATTERNS: readonly RegExp[] = [
  /command not found:\s*([A-Za-z0-9_.+-]+)/,
  /([A-Za-z0-9_.+-]+):\s*command not found/,
  /'([^']+)' is not recognized as an internal or external command/,
  /spawn\s+([A-Za-z0-9_.+\-/\\]+)\s+ENOENT/,
  /ENOENT[^\n]*?['"]([^'"]+)['"]/,
  /([A-Za-z0-9_.+-]+):\s*not found/,
  /No such file or directory:\s*['"]?([A-Za-z0-9_.+\-/\\]+)/,
];

/**
 * Patterns that name what a refused call was refused on, path first: an errno like
 * EACCES is the reason, not the target, so it is only used when nothing else matches.
 */
const DENIED_TARGET_PATTERNS: readonly RegExp[] = [
  /(?:EACCES|EPERM)[^\n]*?['"]([^'"]+)['"]/,
  /[Pp]ermission denied[^\n]*?['"]([^'"]+)['"]/,
  /([^\s:'"]+):\s*[Pp]ermission denied/,
];

/** Errno tokens that must never be reported as the thing a call was denied on. */
const ERRNO_TOKENS = new Set(["EACCES", "EPERM", "ENOENT", "EISDIR", "EROFS"]);

/** True when an error message is about being refused rather than about failing. */
const isDenied = (msg: string) =>
  /permission denied|EACCES|EPERM|operation not permitted|not permitted to|requires approval/i.test(
    msg,
  );

/** First capture from the first matching pattern, trimmed, skipping errno tokens. */
const firstCapture = (msg: string, patterns: readonly RegExp[]) => {
  for (const re of patterns) {
    const m = re.exec(msg);
    const hit = m?.[1]?.trim();
    if (hit && !ERRNO_TOKENS.has(hit)) return hit;
  }
  return undefined;
};

/** One error observation, whatever table it came from. */
interface Observation {
  sid: string;
  tool: string;
  msg: string;
}

/**
 * Errors and failed bash output as one list, without counting a bash failure twice:
 * a failed Bash call already appears in the errors table, so bash output is only
 * used when that table carries no Bash rows at all.
 */
const observationsOf = (errors: readonly ErrorRow[], bash: readonly BashRow[]) => {
  const rows: Observation[] = errors.map((e) => ({
    sid: e.sid,
    tool: e.tool,
    msg: e.msg,
  }));
  if (errors.some((e) => e.tool === "Bash")) return rows;
  for (const b of bash) {
    if (b.err && b.out.length > 0) rows.push({ sid: b.sid, tool: "Bash", msg: b.out });
  }
  return rows;
};

/** Group observations by an extracted target string into count/sessions/sample rows. */
const targetRows = (
  observations: readonly Observation[],
  extract: (o: Observation) => string | undefined,
) => {
  const per = new Map<string, { count: number; sessions: Set<string>; msgs: string[] }>();
  for (const obs of observations) {
    const target = extract(obs);
    if (!target) continue;
    const entry = per.get(target) ?? { count: 0, sessions: new Set<string>(), msgs: [] };
    entry.count += 1;
    entry.sessions.add(obs.sid);
    entry.msgs.push(collapseWs(obs.msg));
    per.set(target, entry);
  }
  return [...per].map(([target, e]) => ({
    target,
    count: e.count,
    sessions: e.sessions.size,
    sampleMsg: clip(shortest(e.msgs), 200),
  }));
};

/** Shortest string, ties broken alphabetically, so samples never vary between runs. */
const shortest = (values: readonly string[]) =>
  [...values].sort((a, b) => a.length - b.length || cmpStr(a, b))[0] ?? "";

/** Errors sharing a normalised signature, ranked by how often they fired. */
const clusterRows = (errors: readonly ErrorRow[], top: number) => {
  const rows: ErrorClusterRow[] = [...groupBy(errors, (e) => e.sig)].map(
    ([sig, group]) => {
      const tools = new Map<string, number>();
      for (const e of group) bump(tools, e.tool);
      return {
        sig: clip(sig, 200),
        count: group.length,
        sessionCount: new Set(group.map((e) => e.sid)).size,
        projects: uniqSorted(group.map((e) => e.proj)).slice(0, 5),
        sampleMsg: clip(shortest(group.map((e) => collapseWs(e.msg))), 240),
        topTool: topKey(tools),
      };
    },
  );
  return ranking(rows, (r) => r.count, (r) => r.sig, top);
};

/**
 * A command re-run within five minutes of failing. Chains are keyed on the exact
 * command (whitespace collapsed) so an unrelated command in between does not break
 * the chain, then reported per family.
 */
const retryRows = (bash: readonly BashRow[], top: number) => {
  const perFamily = new Map<
    string,
    { retries: number[]; sessions: Set<string>; cmds: string[] }
  >();
  for (const [, rows] of groupBy(bash, (r) => r.sid)) {
    const ordered = byTime(rows);
    const consumed = new Set<number>();
    for (let i = 0; i < ordered.length; i += 1) {
      const start = ordered[i];
      if (!start || consumed.has(i) || !start.err) continue;
      const key = cmdKey(start.cmd);
      let lastT = start.t ?? 0;
      let retries = 0;
      for (let j = i + 1; j < ordered.length; j += 1) {
        const next = ordered[j];
        if (!next) continue;
        if ((next.t ?? 0) - lastT > RETRY_WINDOW_SEC) break;
        if (consumed.has(j) || cmdKey(next.cmd) !== key) continue;
        consumed.add(j);
        retries += 1;
        lastT = next.t ?? lastT;
      }
      if (retries === 0) continue;
      const entry = perFamily.get(start.family) ?? {
        retries: [],
        sessions: new Set<string>(),
        cmds: [],
      };
      entry.retries.push(retries);
      entry.sessions.add(start.sid);
      entry.cmds.push(clip(key, 200));
      perFamily.set(start.family, entry);
    }
  }
  const rows: RetryLoopRow[] = [...perFamily].map(([family, e]) => ({
    family,
    count: e.retries.length,
    medianRetries: round(median(e.retries), 2),
    sessions: e.sessions.size,
    sampleCmd: shortest(e.cmds),
  }));
  return ranking(rows, (r) => r.count, (r) => r.family, top);
};

/**
 * What failed: signature clusters, binaries that were never on PATH, refused
 * calls, and commands the agent re-ran after they failed.
 */
export const errorsSection = (
  errors: readonly ErrorRow[],
  bash: readonly BashRow[],
  top: number,
): ErrorsSection => {
  const observations = observationsOf(errors, bash);
  const missing: MissingBinaryRow[] = targetRows(observations, (o) =>
    firstCapture(o.msg, MISSING_BIN_PATTERNS),
  ).map(({ target, ...rest }) => ({ bin: target, ...rest }));
  const denied: PermissionDeniedRow[] = targetRows(observations, (o) =>
    isDenied(o.msg) ? (firstCapture(o.msg, DENIED_TARGET_PATTERNS) ?? o.tool) : undefined,
  );
  return {
    clusters: clusterRows(errors, top),
    missingBinaries: ranking(missing, (r) => r.count, (r) => r.bin, top),
    permissionDenied: ranking(denied, (r) => r.count, (r) => r.target, top),
    retryLoops: retryRows(bash, top),
  };
};

File: skills/inspect-and-debug/harness-doctor/scripts/lib/aggregate_commands.py
"""The `commands` and `errors` sections of the audit.

The Python twin of `lib/aggregate-commands.ts`. These two carry most of the
diagnostic value: `flagFlailing` and `retryLoops` are what catch an agent
groping for the right invocation, so the windowing is spelled out rather than
approximated.
"""

import re

from .stats import (
    bump, by_time, clip, collapse_ws, group_by, median, percentile,
    ranking, round_to, safe_div, top_key, total, uniq_sorted,
)

# Seconds a burst of same-bin commands may span and still count as one flail.
_FLAIL_WINDOW_SEC = 600

# Distinct normalised commands inside one window that make it a flail.
_FLAIL_MIN_DISTINCT = 4

# Seconds after a failure inside which a re-run counts as a retry.
_RETRY_WINDOW_SEC = 300


def _sec_of(row):
    """Duration of a bash call, 0 when the transcript had no result time."""
    return row.get("durSec") or 0


def _cmd_key(cmd):
    """Comparable form of a command: whitespace collapsed, nothing else changed."""
    return collapse_ws(cmd)


def _family_row(family, rows):
    """Roll a set of bash rows sharing a family into one ranked row."""
    secs = [_sec_of(r) for r in rows]
    bins = {}
    for row in rows:
        bump(bins, row.get("bin", ""))
    return {
        "family": family,
        "bin": top_key(bins),
        "count": len(rows),
        "totalSec": round_to(total(secs), 1),
        "medianSec": round_to(median(secs), 2),
        "p95Sec": round_to(percentile(secs, 0.95), 2),
        "errRate": round_to(safe_div(len([r for r in rows if r.get("err")]), len(rows)), 4),
    }


def command_families(bash):
    """Command families ranked rows, reused per project by the projects section."""
    grouped = group_by(bash, lambda r: r.get("family", ""))
    rows = [_family_row(family, items) for family, items in grouped.items()]
    return sorted(rows, key=lambda r: r["family"])


def _repeat_rows(bash, top):
    """Identical command re-run inside one session: everything after the first."""
    per_cmd = {}
    for _, rows in group_by(bash, lambda r: "%s %s" % (r["sid"], _cmd_key(r.get("cmd", "")))).items():
        if len(rows) < 2:
            continue
        ordered = sorted(rows, key=lambda r: r.get("t") or 0)
        first = ordered[0]
        key = _cmd_key(first.get("cmd", ""))
        entry = per_cmd.setdefault(key, {
            "family": first.get("family", ""), "cmd": clip(key, 200),
            "occurrences": 0, "sessions": set(), "wastedSec": 0,
        })
        entry["occurrences"] += len(ordered) - 1
        entry["sessions"].add(first["sid"])
        entry["wastedSec"] += total(ordered[1:], _sec_of)
    rows = [{
        "family": e["family"], "cmd": e["cmd"], "occurrences": e["occurrences"],
        "sessions": len(e["sessions"]), "wastedSec": round_to(e["wastedSec"], 1),
    } for e in per_cmd.values()]
    return ranking(rows, lambda r: r["occurrences"], lambda r: r["cmd"], top)


def _bursts_of(rows):
    """Split time-ordered rows into bursts opened by the first row outside the window."""
    out = []
    current = None
    opened_at = 0
    for row in rows:
        t = row.get("t") or 0
        if current is None or t - opened_at > _FLAIL_WINDOW_SEC:
            current = {"rows": [], "families": set()}
            opened_at = t
            out.append(current)
        current["rows"].append(row)
        current["families"].add(row.get("family", ""))
    return out


def _flailing_rows(bash, top):
    """Same bin, four or more distinct commands inside ten minutes of one session."""
    per_bin = {}
    for key, rows in group_by(bash, lambda r: "%s %s" % (r["sid"], r.get("bin", ""))).items():
        sid, _, binname = key.partition(" ")
        for burst in _bursts_of(by_time(rows)):
            if len(burst["families"]) < _FLAIL_MIN_DISTINCT:
                continue
            entry = per_bin.setdefault(binname, {
                "sessions": set(), "occurrences": 0, "totalSec": 0, "cmds": set(),
            })
            entry["sessions"].add(sid)
            entry["occurrences"] += len(burst["families"])
            entry["totalSec"] += total(burst["rows"], _sec_of)
            for row in burst["rows"]:
                entry["cmds"].add(clip(_cmd_key(row.get("cmd", "")), 160))
    rows = [{
        "bin": binname, "sessions": len(e["sessions"]), "occurrences": e["occurrences"],
        "totalSec": round_to(e["totalSec"], 1), "sampleCmds": uniq_sorted(e["cmds"])[:4],
    } for binname, e in per_bin.items()]
    return ranking(rows, lambda r: r["occurrences"], lambda r: r["bin"], top)


def commands_section(bash, top):
    """Bash cost, ranked and diagnosed."""
    families = command_families(bash)
    return {
        "byTotalSec": ranking(families, lambda r: r["totalSec"], lambda r: r["family"], top),
        "byCount": ranking(families, lambda r: r["count"], lambda r: r["family"], top),
        "repeatsInSession": _repeat_rows(bash, top),
        "flagFlailing": _flailing_rows(bash, top),
    }


# Patterns that name a binary the shell could not find, most specific first.
_MISSING_BIN_PATTERNS = (
    re.compile(r"command not found:\s*([A-Za-z0-9_.+-]+)"),
    re.compile(r"([A-Za-z0-9_.+-]+):\s*command not found"),
    re.compile(r"'([^']+)' is not recognized as an internal or external command"),
    re.compile(r"spawn\s+([A-Za-z0-9_.+\-/\\]+)\s+ENOENT"),
    re.compile(r"ENOENT[^\n]*?['\"]([^'\"]+)['\"]"),
    re.compile(r"([A-Za-z0-9_.+-]+):\s*not found"),
    re.compile(r"No such file or directory:\s*['\"]?([A-Za-z0-9_.+\-/\\]+)"),
)

# Patterns that name what a refused call was refused on, path first.
_DENIED_TARGET_PATTERNS = (
    re.compile(r"(?:EACCES|EPERM)[^\n]*?['\"]([^'\"]+)['\"]"),
    re.compile(r"[Pp]ermission denied[^\n]*?['\"]([^'\"]+)['\"]"),
    re.compile(r"([^\s:'\"]+):\s*[Pp]ermission denied"),
)

# Errno tokens that must never be reported as the thing a call was denied on.
_ERRNO_TOKENS = frozenset(["EACCES", "EPERM", "ENOENT", "EISDIR", "EROFS"])

_DENIED = re.compile(
    r"permission denied|EACCES|EPERM|operation not permitted|not permitted to|requires approval",
    re.I,
)


def _first_capture(msg, patterns):
    """First capture from the first matching pattern, skipping errno tokens."""
    for pattern in patterns:
        m = pattern.search(msg or "")
        hit = m.group(1).strip() if m and m.group(1) else None
        if hit and hit not in _ERRNO_TOKENS:
            return hit
    return None


def _observations_of(errors, bash):
    """Errors and failed bash output as one list, without counting a failure twice.

    A failed Bash call already appears in the errors table, so bash output is
    only used when that table carries no Bash rows at all.
    """
    rows = [{"sid": e["sid"], "tool": e.get("tool", "unknown"), "msg": e.get("msg", "")} for e in errors]
    if any(e.get("tool") == "Bash" for e in errors):
        return rows
    for b in bash:
        if b.get("err") and b.get("out"):
            rows.append({"sid": b["sid"], "tool": "Bash", "msg": b["out"]})
    return rows


def _shortest(values):
    """Shortest string, ties alphabetical, so samples never vary between runs."""
    ordered = sorted(values, key=lambda v: (len(v), v))
    return ordered[0] if ordered else ""


def _target_rows(observations, extract):
    """Group observations by an extracted target into count/sessions/sample rows."""
    per = {}
    for obs in observations:
        target = extract(obs)
        if not target:
            continue
        entry = per.setdefault(target, {"count": 0, "sessions": set(), "msgs": []})
        entry["count"] += 1
        entry["sessions"].add(obs["sid"])
        entry["msgs"].append(collapse_ws(obs["msg"]))
    return [{
        "target": target, "count": e["count"], "sessions": len(e["sessions"]),
        "sampleMsg": clip(_shortest(e["msgs"]), 200),
    } for target, e in per.items()]


def _cluster_rows(errors, top):
    """Errors sharing a normalised signature, ranked by how often they fired."""
    rows = []
    for sig, group in group_by(errors, lambda e: e.get("sig", "")).items():
        tools = {}
        for e in group:
            bump(tools, e.get("tool", "unknown"))
        rows.append({
            "sig": clip(sig, 200),
            "count": len(group),
            "sessionCount": len({e["sid"] for e in group}),
            "projects": uniq_sorted(e["proj"] for e in group)[:5],
            "sampleMsg": clip(_shortest([collapse_ws(e.get("msg", "")) for e in group]), 240),
            "topTool": top_key(tools),
        })
    return ranking(rows, lambda r: r["count"], lambda r: r["sig"], top)


def _retry_rows(bash, top):
    """A command re-run within five minutes of failing, reported per family."""
    per_family = {}
    for _, rows in group_by(bash, lambda r: r["sid"]).items():
        ordered = by_time(rows)
        consumed = set()
        for i, start in enumerate(ordered):
            if i in consumed or not start.get("err"):
                continue
            key = _cmd_key(start.get("cmd", ""))
            last_t = start.get("t") or 0
            retries = 0
            for j in range(i + 1, len(ordered)):
                nxt = ordered[j]
                if (nxt.get("t") or 0) - last_t > _RETRY_WINDOW_SEC:
                    break
                if j in consumed or _cmd_key(nxt.get("cmd", "")) != key:
                    continue
                consumed.add(j)
                retries += 1
                last_t = nxt.get("t") or last_t
            if retries == 0:
                continue
            entry = per_family.setdefault(start.get("family", ""), {
                "retries": [], "sessions": set(), "cmds": [],
            })
            entry["retries"].append(retries)
            entry["sessions"].add(start["sid"])
            entry["cmds"].append(clip(key, 200))
    rows = [{
        "family": family, "count": len(e["retries"]),
        "medianRetries": round_to(median(e["retries"]), 2),
        "sessions": len(e["sessions"]), "sampleCmd": _shortest(e["cmds"]),
    } for family, e in per_family.items()]
    return ranking(rows, lambda r: r["count"], lambda r: r["family"], top)


def errors_section(errors, bash, top):
    """Signature clusters, missing binaries, refused calls, and retry loops."""
    observations = _observations_of(errors, bash)
    missing = [
        {"bin": r.pop("target"), **r}
        for r in _target_rows(observations, lambda o: _first_capture(o["msg"], _MISSING_BIN_PATTERNS))
    ]
    denied = _target_rows(
        observations,
        lambda o: (_first_capture(o["msg"], _DENIED_TARGET_PATTERNS) or o["tool"])
        if _DENIED.search(o["msg"] or "") else None,
    )
    return {
        "clusters": _cluster_rows(errors, top),
        "missingBinaries": ranking(missing, lambda r: r["count"], lambda r: r["bin"], top),
        "permissionDenied": ranking(denied, lambda r: r["count"], lambda r: r["target"], top),
        "retryLoops": _retry_rows(bash, top),
    }
